updater/update.go

```diff
@@ -229,3 +226,71 @@ func CopyFile(src, dst string) (int64, error) {
 	nBytes, err := io.Copy(destination, source)
 	return nBytes, err
 }
+
+// candidateWysFileMatchesFailedInstallWysFile considers whether the WYS file read from the candidateWysFileReader, which
+// corresponds to an update that is a candidate for further processing, matches the locally saved copy of the last WYS file
+// we tried to process and install.  It will return false if we should further process the candidate WYS file, and true otherwise.
+func candidateWysFileMatchesFailedInstallWysFile(candidateWysFileReader io.Reader) bool {
+	// we expect this failed sentinel file to exist iff a prior update/installation failed
+	// it should be a copy of the WYS file that triggered this aforementioned update
+	installFailedSentinelFilePath := filepath.Join(GetExeDir(), INSTALL_FAILED_SENTINAL_WYS_FILE_NAME)
+	if !fileExists(installFailedSentinelFilePath) {
+		return false
+	}
+
+	sentinelWysFileHash, err := GenerateSHA1HashFromFilePath(installFailedSentinelFilePath)
+	if err != nil {
+		return false
+	}
+
+	candidateWysFileHash, err := GenerateSHA1HashFromReader(candidateWysFileReader)
+	if err != nil {
+		return false
+	}
+
+	return bytes.Equal(sentinelWysFileHash, candidateWysFileHash)
+}
+
+// A CandidateUpdateContext represents all the data necessary to define and generate a request for an [agent] update
+type CandidateUpdateContext struct {
+	ConfigIUC               ConfigIUC
+	CandidateWysFileContent bytes.Buffer
+	ConfigWYS               ConfigWYS
+}
+
+// GenerateCandidateUpdateContext returns a populated ctx if all the prerequisites to generate one are present.
+// ctx will always be zero value when err is not nil.
+func GenerateCandidateUpdateContext(args Args, wyFileParser Infoer) (ctx CandidateUpdateContext, err error) {
+	// parse the WYC file to get the update site, installed version, etc.
+	wycFilePath := args.Cdata
+	iuc, err := wyFileParser.ParseWYC(wycFilePath)
+	if nil != err {
+		err = fmt.Errorf("error reading WYC file: %s; %v", wycFilePath, err)
+		return ctx, err
+	}
+
+	urls := iuc.GetWYSURLs(args)
+
+	var candidateWysFileContents bytes.Buffer
+	err = DownloadFileToWriter(urls, &candidateWysFileContents)
+	if nil != err {
+		return ctx, err
+	}
+
+	candidateWysFileReader := bytes.NewReader(candidateWysFileContents.Bytes())
+	wys, err := wyFileParser.ParseWYSFromReader(candidateWysFileReader, int64(candidateWysFileContents.Len()))
+	if nil != err {
+		err = fmt.Errorf("error parsing downloaded candidate WYS file; %v", err)
```

Comment:
```suggestion
		err = fmt.Errorf("error parsing downloaded candidate WYS file; %w", err)
```
Ok. I'm going to stop suggesting these now, but the point still stands. 😄

Reply:
😄 yeah I'll review all the things I touched and shift to %w for errors. Thx for the insight.